### PR TITLE
Refactor attestation and registration

### PR DESCRIPTION
**Pull request by jasl**
This is a bunch refactors about worker registration and attestation:

- Remove AttestationValidator type because in real world we will support multiple attestation providers (IAS, DCAP, etc.)
- Refactor attestation validate function to support multiple attestation providers
- Add OptOut attestation provider for dev & test chain (production chain should disable this)
- Store attestation provider on WorkerInfo
- `messaging::WorkerInfo` also includes `attestation_provider`

### CODE REVIEW

**Comment by kvinwang** [comment]
> * Remove AttestationValidator type because in real world we will support multiple attestation providers (IAS, DCAP, etc.)

Is it better to define a `struct MultiValidator` and `impl AttestationValidator for MultiValidator`? The return type of `fn validate` might need to be changed.

**Comment by jasl** [comment]
> > * Remove AttestationValidator type because in real world we will support multiple attestation providers (IAS, DCAP, etc.)
> 
> Is it better to define a `struct MultiValidator` and `impl AttestationValidator for MultiValidator`? The return type of `fn validate` might need to be changed.

That was my plan, but, because of we need to support all kinds of attestation methods at the same time, it seems there's no benefit for extract a struct.

File: crates/phactory/src/lib.rs [hunk 1]
```diff
@@ -203,10 +205,12 @@ impl<Platform: pal::Platform> Phactory<Platform> {
         }
 
         self.args = args;
+        self.attestation_provider = self.args.attestation_provider.clone();
     }
 
     pub fn set_args(&mut self, args: InitArgs) {
         self.args = args;
+        self.attestation_provider = self.args.attestation_provider.clone();
```
Comment (kvinwang) [inline crates/phactory/src/lib.rs:213]: The `set_args` is called immediately after restored from checkpoint. So the `attestation_provider` should keep the value from checkpoint. Infact, any parameters that could not be changed over a restoring should not be put in the command line args. Better put it in the `InitRuntime` request?

File: crates/phactory/src/prpc_service.rs [hunk 2]
```diff
@@ -412,13 +413,8 @@ impl<Platform: pal::Platform + Serialize + DeserializeOwned> Phactory<Platform>
...
                     version: 1,
-                    provider: "SGX".to_string(),
-                    payload: Some(pb::AttestationReport {
-                        report: attn_report,
-                        signature: base64::decode(sig).map_err(from_display)?,
-                        signing_cert: base64::decode_config(cert, base64::STANDARD)
-                            .map_err(from_display)?,
-                    }),
+                    provider: self.attestation_provider.clone(),
+                    payload: encoded_report,
```
Comment (kvinwang) [inline crates/phactory/src/prpc_service.rs:417]: This is a breaking change which make a new version of `pherry` or `prb` won't talk to old `pruntime`s. Better create a new field, like `payload_v2`?

File: crates/phala-types/src/lib.rs [hunk 3]
```diff
@@ -644,6 +667,7 @@ pub struct WorkerInfo<BlockNumber> {
     // performance
     pub score: Option<Score>,
     // confidence-level
+    pub attestation_provider: AttestationProvider,
```
Comment (kvinwang) [inline crates/phala-types/src/lib.rs:670]: Remove the comment.

File: pallets/phala/src/registry.rs [hunk 4]
```diff
@@ -45,7 +45,10 @@ pub mod pallet {
 		type Currency: Currency<Self::AccountId>;
 
 		type UnixTime: UnixTime;
-		type AttestationValidator: AttestationValidator;
+
+		/// Enable OptOut Attestation, SHOULD NOT SET FALSE ON PRODUCTION !!!
```
Comment (kvinwang) [inline pallets/phala/src/registry.rs:49]:
```suggestion
		/// Enable OptOut Attestation, SHOULD BE SET TO FALSE ON PRODUCTION !!!
```

File: crates/phactory/src/lib.rs (hunk 1 above)
Comment (jasl) [reply crates/phactory/src/lib.rs:213]: Personal I don't like add this to `InitRuntime`, but I saw `skip_ra` already there...

**Comment by jasl** [comment]
close this in favor of https://github.com/Phala-Network/phala-blockchain/pull/768

**Pull request closed without merging**